pallets/parachain-staking/src/lib.rs

```diff
@@ -839,25 +837,63 @@ pub mod pallet {
 		///   `SelectedCandidates` storage in `select_top_candidates` which in
 		///   return depends on the number of `MaxSelectedCandidates` (N).
-		/// - For each N, we read `CollatorState` from the storage.
+		/// - For each N, we read `CandidatePool` from the storage.
 		/// ---------
-		/// Weight: O(N) where N is `MaxSelectedCandidates` bounded by
-		/// `MaxTopCandidates`
-		/// - Reads: MaxSelectedCandidates, TopCandidates, N * CollatorState
+		/// Weight: O(N + D) where N is `MaxSelectedCandidates` bounded by
+		/// `MaxTopCandidates` and D is the number of delegators of a
+		/// candidate bounded by `MaxDelegatorsPerCollator`.
+		/// - Reads: MaxSelectedCandidates, TopCandidates, N * CandidatePool
 		/// - Writes: MaxSelectedCandidates
 		/// # </weight>
-		#[pallet::weight(<T as pallet::Config>::WeightInfo::set_max_selected_candidates(*new, (*new).saturating_mul(T::MaxDelegatorsPerCollator::get())))]
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::set_max_selected_candidates(
+			*new,
+			T::MaxDelegatorsPerCollator::get()
+		))]
 		pub fn set_max_selected_candidates(origin: OriginFor<T>, new: u32) -> DispatchResultWithPostInfo {
 			ensure_root(origin)?;
 			ensure!(new >= T::MinCollators::get(), Error::<T>::CannotSetBelowMin);
-			let old = <MaxSelectedCandidates<T>>::get();
+			ensure!(new <= T::MaxTopCandidates::get(), Error::<T>::CannotSetAboveMax);
+			let old = MaxSelectedCandidates::<T>::get();
 
 			// *** No Fail beyond this point ***
 
-			<MaxSelectedCandidates<T>>::put(new);
-
-			// update candidates for next round
-			let (num_collators, num_delegators, _, _) = Self::update_total_stake();
+			MaxSelectedCandidates::<T>::put(new);
+
+			// Update total amount at stake for new top collators and their delegators
+			let (num_collators, num_delegators) = match new {
+				x if x > old => {
+					let candidates = TopCandidates::<T>::get();
+					let num_new_collators: usize = candidates
+						.len()
+						.min(new.saturated_into())
+						.saturating_sub(old.saturated_into());
+					let (col_stake, del_stake, num_delegators) =
+						// FIXME: Check for more simple solution to take slice [old + 1, ..., num_new]
+						candidates.into_iter().take(new.saturated_into()).rev().take(num_new_collators).fold(
+							(BalanceOf::<T>::zero(), BalanceOf::<T>::zero(), 0u32),
+							|(col_stake, del_stake, num_delegators), candidate| {
+								if let Some(state) = CandidatePool::<T>::get(&candidate.owner) {
+									(
+										col_stake.saturating_add(state.stake),
+										del_stake.saturating_add(state.total - state.stake),
+										num_delegators.max(state.delegators.len().saturated_into()),
+									)
+								} else {
+									(col_stake, del_stake, num_delegators)
+								}
+							},
+						);
+
+					TotalCollatorStake::<T>::mutate(|total| {
+						total.collators = total.collators.saturating_add(col_stake);
+						total.delegators = total.delegators.saturating_add(del_stake);
+					});
```

Comment:
we don't read from the old total. So i would just use `set`?

Reply:
We do read from the old total by incrementing the existing values?!

Reply:
completely overlooked that. 😂